src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.KeepAlive.cs

```diff
@@ -0,0 +1,306 @@
+        }
+
+        private void KeepAlivePingHeartBeat()
+        {
+            Debug.Assert(_keepAlivePingState != null);
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+            bool shouldSendPing = false;
+            long pingPayload = -1;
+
+            try
+            {
+                lock (StateUpdateLock)
+                {
+                    if (_keepAlivePingState.Exception is not null)
+                    {
+                        if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"KeepAlive already faulted, skipping... (exception: {_keepAlivePingState.Exception.Message})");
+                        return;
+                    }
+
+                    long now = Environment.TickCount64;
+
+                    if (_keepAlivePingState.PingSent)
+                    {
+                        if (Environment.TickCount64 > _keepAlivePingState.PingTimeoutTimestamp)
+                        {
+                            if (NetEventSource.Log.IsEnabled())
+                            {
+                                NetEventSource.Trace(this, $"Keep-alive ping timed out after {_keepAlivePingState.TimeoutMs}ms. Expected pong with payload {_keepAlivePingState.PingPayload}");
+                            }
+
+                            Exception exc = ExceptionDispatchInfo.SetCurrentStackTrace(
+                                new WebSocketException(WebSocketError.Faulted, SR.net_Websockets_KeepAlivePingTimeout));
+
+                            _keepAlivePingState.OnKeepAliveFaultedCore(exc); // we are holding the lock
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        if (Environment.TickCount64 > _keepAlivePingState.NextPingRequestTimestamp)
+                        {
+                            _keepAlivePingState.OnNextPingRequestCore(); // we are holding the lock
+                            shouldSendPing = true;
+                            pingPayload = _keepAlivePingState.PingPayload;
+                        }
+                    }
+                }
+
+                if (shouldSendPing)
+                {
+                    Observe(
+                        SendPingAsync(pingPayload));
+                }
+            }
+            catch (Exception e)
+            {
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.TraceException(this, e);
+
+                _keepAlivePingState.OnKeepAliveFaulted(e);
+            }
+        }
+
+        private async ValueTask SendPingAsync(long pingPayload)
+        {
+            Debug.Assert(_keepAlivePingState != null);
+
+            byte[] pingPayloadBuffer = ArrayPool<byte>.Shared.Rent(sizeof(long));
+            BinaryPrimitives.WriteInt64BigEndian(pingPayloadBuffer, pingPayload);
```

Comment:
This is only 8 bytes presumably needed only once every few seconds at most. Is renting worthwhile?